EIPS/eip-1654.md

```diff
@@ -0,0 +1,107 @@
+2. Dapp client sends the signature to the dapp's backend component, along with the wallet address to be authenticated with. The address may be obtained via [`eth_accounts`](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_accounts).
+3. Dapp backend recovers a public key from the signature.
+4. Dapp backend checks if the recovered key has authorized control over the provided wallet address under the assumption it could represent an external wallet OR a contract wallet. For the case of a contract wallet, it MUST call `isValidSignature()` and expect the value `0x1626ba7e` to determine whether the entity who signed the challenge has authorized control over the wallet.
+5. The result of the authorized control check is returned as the result of the authentication and the flow is complete.
+
+A challenge message SHOULD contain a random component. This will reduce the risk of replay attacks.
+
+A challenge message SHOULD be generated by the dapp backend AND not get sent back as input from the dapp client, but be persisted in the backend for at least the entirety of the authentication process. This will remove the risk of accepting forged challenges.
+
+The following algorithm MAY be used by dapp backend for authenticating users with personal signed messages:
+
+```
+FUNCTION IsAuthorizedSigner(challengeString, signature, walletAddress) RETURNS (success)
+
+  SET personalChallengeHash to the hash of: challengeString prepended with `"\x19Ethereum Signed Message:\n" + len(challengeString)`
+
+  SET recoveredKey to the public key recovered from signature and personalChallengeHash
+
+  SET recoveredAddress to the address corresponding with recoveredKey
+
+  // try external wallet
+  IF walletAddress EQUALS recoveredAddress
+    RETURN true
+  END IF
+
+  SET challengeHash to the hash of: challengeString . We send just a regular Keccak256 hash, which then the smart contract hashes ontop to an erc191 hash.
+
+  SET contractResult to the result of calling isValidSignature(challengeHash, signature) on the contract at walletAddress 
+
+  IF contractResult EQUALS 0x1626ba7e
+    RETURN true
+  ELSE
+    RETURN false
+  END IF
+
+END FUNCTION
+```
+
+### Wallet
+
+#### External wallet
+Any software agents managing external wallets are not required to make any changes to continue to work with this process.
+
+#### Contract wallet
+
+##### Contract
+
+Must implement latest [EIP-1271](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1271.md) with `MAGICVALUE = 0x1626ba7e`.
+
+##### User agent
+
+A user agent intended to work with the contract MUST generate signatures over a EIP-191 hash of a regular Keccak256 hash of the challenge message.
+
+## Rationale
+[EIP-1271](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1271.md) has done a great work with standardising the signature validation method for contracts. This proposal builds on top of it a process for dapp-wallet authentication. 
```

Comment:
```suggestion
TBD
```
Rationale is a section for explaining why you made some specific choices that you made.  This statement seems closer to a motivation... though it doesn't really add any value so I recommend deleting it entirely.